### PR TITLE
Fix tab completion for unlocalized about* topics

### PR DESCRIPTION
**Pull request by MartinGC94**
Fixes tab completion for help topics on systems without localized help by adding en-US as a fallback.
`Get-Help About<Tab>`

## PR Context

Get-Help will automatically fall back to en-US help topics so the completion for it should do the same.

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [X] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [X] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [X] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs [hunk 1]
```diff
@@ -6270,35 +6270,37 @@ private static string GetNamespaceToRemove(CompletionContext context, TypeComple
...
                 var wildcardPattern = WildcardPattern.Get(wordToComplete, WildcardOptions.IgnoreCase);
 
                 foreach (var dir in searchPaths)
                 {
-                    foreach (var file in Directory.EnumerateFiles(dir))
+                    if (Directory.Exists(dir))
                     {
-                        if (wildcardPattern.IsMatch(Path.GetFileName(file)))
+                        var currentDir = new DirectoryInfo(dir);
+                        foreach (var file in currentDir.EnumerateFiles("about_*.help.txt"))
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:6298]: If you are already using DirectoryInfo:
```c#
                    var currentDir = new DirectoryInfo(dir);
                    if (currentDir.Exists)
                    {
                        foreach (var file in currentDir.EnumerateFiles("about_*.help.txt"))
```

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs [hunk 2]
```diff
@@ -6307,30 +6309,9 @@ internal static List<CompletionResult> CompleteHelpTopics(CompletionContext cont
...
-                        // All topic files are ending with ".help.txt"
-                        var completionText = fileName.Substring(0, fileName.Length - 9);
-                        results.Add(new CompletionResult(completionText));
-                    }
-                    catch (Exception)
-                    {
-                        continue;
-                    }
-                }
+                results.Add(new CompletionResult(topic));
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:6314]: Please move in previous cycle (line 6302).

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs (hunk 1 above)
Comment (MartinGC94) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs]: Good idea, I wrongly assumed that new DirectoryInfo would throw an error if the folder didn't exist.

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs [hunk 3]
```diff
@@ -6270,69 +6270,49 @@ private static string GetNamespaceToRemove(CompletionContext context, TypeComple
...
-
-            var wordToComplete = context.WordToComplete + "*";
-            var topicPattern = WildcardPattern.Get("about_*.help.txt", WildcardOptions.IgnoreCase);
-            List<string> files = new List<string>();
+            string userHelpDir = HelpUtils.GetUserHomeHelpSearchPath();
+            string appHelpDir = Utils.GetApplicationBase(Utils.DefaultPowerShellShellID);
+            string currentCulture = CultureInfo.CurrentCulture.Name;
+
+            //search for help files for the current culture + en-US as fallback
+            var searchPaths = new string[4]
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:6278]:
```suggestion
            var searchPaths = new string[]
```

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs [hunk 4]
```diff
@@ -6270,69 +6270,49 @@ private static string GetNamespaceToRemove(CompletionContext context, TypeComple
...
+            { 
+                Path.Combine(userHelpDir, currentCulture),
+                Path.Combine(userHelpDir, "en-US"),
+                Path.Combine(appHelpDir, currentCulture),
+                Path.Combine(appHelpDir, "en-US")
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:6283]: For discussion:
```suggestion
                Path.Combine(userHelpDir, currentCulture),
                Path.Combine(appHelpDir, currentCulture),
                Path.Combine(userHelpDir, "en-US"),
                Path.Combine(appHelpDir, "en-US")
```
Maybe users want to get topics in their language first?

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs [hunk 5]
```diff
@@ -6270,69 +6270,49 @@ private static string GetNamespaceToRemove(CompletionContext context, TypeComple
...
                         {
-                            files.Add(file);
+                            if (wildcardPattern.IsMatch(file.Name))
+                            {
+                                string topicName = file.Name.Substring(0, file.Name.LastIndexOf(".help.txt"));
+                                bool notDuplicate = foundTopics.Add(topicName);
+                                if (notDuplicate)
+                                {
+                                    results.Add(new CompletionResult(topicName));
+                                }
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:6307]: Removing duplications changes current behavior. For the PR this looks as unrelated change. It seems we have an issue for the scenario and we shouldn't remove duplicates but allow users to select. I suggest revert the change in the PR.

Comment (MartinGC94) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:6307]: Duplicates serve no purpose, if we don't filter them out here then PSReadline will do it for us and if that doesn't do it then the user will simply end up with 2 identical tab completion results and probably assume something is broken.  
Get-Help would have to be updated to support specifying a different language and I think that's a good idea, but I think it should be done with another parameter, not by cluttering up tab completion suggestions with near identical suggestions.

Comment (iSazonov) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:6307]: See #11205. I suggest to address the issue in separate PR and keep current behavior in the PR.

**Review by adityapatwardhan — CHANGES_REQUESTED** [review]

**Review by iSazonov — APPROVED** [review]

**Review by adityapatwardhan — APPROVED** [review]

**Comment by adityapatwardhan** [comment]
@MartinGC94 Thank you for your contribution!
